pkg/operators/ebpf/ebpf.go

```diff
@@ -544,13 +544,45 @@ func (i *ebpfInstance) Start(gadgetCtx operators.GadgetContext) error {
 			i.links = append(i.links, l)
 		}
 
-		// we need to store links to iterators on a separated list because we need them to run the programs.
+		// We need to store links to iterators because we need them to run the programs
 		if p.Type == ebpf.Tracing && strings.HasPrefix(p.SectionName, iterPrefix) {
 			lIter, ok := l.(*link.Iter)
 			if !ok {
 				return fmt.Errorf("link is not an iterator")
 			}
-			i.linksSnapshotters = append(i.linksSnapshotters, &linkSnapshotter{link: lIter, typ: p.AttachTo})
+
+			for _, snapshotter := range i.snapshotters {
+				if _, ok := snapshotter.iterators[progName]; ok {
+					snapshotter.links[progName] = &linkSnapshotter{
+						link: lIter,
+						typ:  p.AttachTo,
+					}
+					continue programs
+				}
+			}
+
+			// We have an iterator that was not associated with any snapshotter.
+			// If we have only one snapshotter and it was not associated with any
+			// iterator, we can associate all iterators with it.
```

Comment:
This sounds like magic that could lead to unwanted behavior and it might be difficult to understand for gadget authors. I'd much more prefer to have this setup explicitly (preferably using a macro, hehe).

Reply:
Let's discuss it in [this other thread](https://github.com/inspektor-gadget/inspektor-gadget/pull/2637#discussion_r1542633847).